internal/helm_client.go

```diff
@@ -255,21 +259,63 @@ func (c *HelmClient) InstallHelmRelease(ctx context.Context, restConfig *rest.Co
 
 // newDefaultRegistryClient creates registry client object with default config which can be used to install/upgrade helm charts.
 func newDefaultRegistryClient(credentialsPath string, enableCache bool, caFilePath string, insecureSkipTLSVerify bool) (*registry.Client, error) {
-	if caFilePath == "" && !insecureSkipTLSVerify {
-		opts := []registry.ClientOption{
-			registry.ClientOptDebug(true),
-			registry.ClientOptEnableCache(enableCache),
-			registry.ClientOptWriter(os.Stderr),
-		}
-		if credentialsPath != "" {
-			// Create a new registry client with credentials
-			opts = append(opts, registry.ClientOptCredentialsFile(credentialsPath))
+	opts := []registry.ClientOption{
+		registry.ClientOptDebug(true),
+		registry.ClientOptEnableCache(enableCache),
+		registry.ClientOptWriter(os.Stderr),
+	}
+	if credentialsPath != "" {
+		// Create a new registry client with credentials
+		opts = append(opts, registry.ClientOptCredentialsFile(credentialsPath))
+	}
+
+	if caFilePath != "" || insecureSkipTLSVerify {
+		tlsConf, err := newClientTLS(caFilePath, insecureSkipTLSVerify)
+		if err != nil {
+			return nil, fmt.Errorf("can't create TLS config for client: %s", err)
 		}
+		opts = append(opts, registry.ClientOptHTTPClient(&http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: tlsConf,
+				Proxy:           http.ProxyFromEnvironment,
+				// This registry client is not reused and is discarded after a single reconciliation
+				// loop. Limit how long can be the idle connection open. Otherwise its possible that
+				// a registry server that keeps the connection open for a long time could result in
+				// connections being openend in the controller for a long time.
+				IdleConnTimeout: 1 * time.Second,
+			},
+		}),
+		)
+	}
+	return registry.NewClient(opts...)
+}
 
-		return registry.NewClient(opts...)
+func newClientTLS(caFile string, insecureSkipTLSverify bool) (*tls.Config, error) {
+	config := tls.Config{
+		InsecureSkipVerify: insecureSkipTLSverify,
 	}
 
-	return registry.NewRegistryClientWithTLS(os.Stderr, "", "", caFilePath, insecureSkipTLSVerify, credentialsPath, true)
+	if caFile != "" {
+		cp, err := certPoolFromFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+		config.RootCAs = cp
+	}
+
+	return &config, nil
+}
+
+func certPoolFromFile(filename string) (*x509.CertPool, error) {
```

Comment:
Aside from the refactor, this part is new. What will change from using this approach specifically?

Reply:
These functions were added to create http TLS configuration based on the provided caFilePath and insecure TLS flags. The refactor is removing use of helm utility `NewRegistryClientWithTLS` function - https://github.com/helm/helm/blob/v3.16.4/pkg/registry/util.go#L143-L166 . The function calls `tlsutils.NewClientTLS` which is in helm internal package. I've copied bits that are relevant for CAAPH configuration options.